databuilder/query_engines/mssql.py

```diff
@@ -2,12 +2,17 @@
 
 from databuilder import sqlalchemy_types
 from databuilder.query_engines.base_sql import BaseSQLQueryEngine
-from databuilder.query_engines.mssql_dialect import MSSQLDialect
+from databuilder.query_engines.mssql_dialect import MSSQLDialect, SelectStarInto
+from databuilder.sqlalchemy_utils import GeneratedTable
 
 
 class MSSQLQueryEngine(BaseSQLQueryEngine):
     sqlalchemy_dialect = MSSQLDialect
 
+    # The `#` prefix is an MSSQL-ism which automatically makes the tables session-scoped
+    # temporary tables
+    intermediate_table_prefix = "#tmp_"
```

Comment:
It makes me a little uncomfortable that we subclass the _entire_ base engine and then override arbitrary bits of it. So reading the base class there is no way of knowing whether the thing you are looking at is overridden in some subclass. I suppose we might have some sort of "dialect" mechanism where there is a "base" class that contains only logic that doesn't change and delegates to a specializing object that contains all the variation.

I don't feel strongly about this. Just a thought.

Reply:
Yeah, this has been niggling at me as well but I sort of pushed the worry aside because I didn't have a clear alternative proposal. But you're quite right that it makes the control flow impossible to follow, and it's only going to get worse as we add further sorts of variation.

I think having some kind of dialect object is probably the way to go. Although that does mean yet another layer of indirection.

An alternative would be, for any behaviour where we want variation, to refuse to implement default behaviour in the base class and to require each subclass to explicitly implement its own behaviour. So for instance `BaseSQLQueryEngine` would just have
```py
def reify_query(self, query):
    raise NotImplementedError()
```

That could get quite verbose and repetitive quickly though.

Reply:
Yes, I think it's reasonable to have an inheritance hierarchy with overridden methods to remove duplication. But it would be helpful to have a signal for which methods _might_ be overridden.